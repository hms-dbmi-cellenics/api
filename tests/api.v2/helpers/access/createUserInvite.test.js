// Disabled ts because it doesn't recognize jest mocks
// @ts-nocheck
const UserAccess = require('../../../../src/api.v2/model/UserAccess');

const createUserInvite = require('../../../../src/api.v2/helpers/access/createUserInvite');
const AccessRole = require('../../../../src/utils/enums/AccessRole');
const sendEmail = require('../../../../src/utils/sendEmail');
const { getAwsUserAttributes } = require('../../../../src/utils/aws/user');

jest.mock('../../../../src/utils/aws/user');
jest.mock('../../../../src/api.v2/model/UserAccess');
jest.mock('../../../../src/utils/sendEmail');

const mockUserAccess = {
  grantAccess: jest.fn(),
  addToInviteAccess: jest.fn(),
};

UserAccess.mockReturnValue(mockUserAccess);

const mockExperimentId = 'experimentId';
const mockInvitedUserEmail = 'invited@example.com';
const mockInviterUser = { email: 'inviter@example.com' };
const mockRole = AccessRole.EXPLORER;

process.env.DOMAIN_NAME = 'localhost.test';

describe('creatUserInvite', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  it('Grants access if user is already registered and sends an email', async () => {
    getAwsUserAttributes.mockImplementationOnce(() => [
      { Name: 'sub', Value: 'mock-user-id' },
      { Name: 'email_verified', Value: 'true' },
      { Name: 'name', Value: 'Mock Invited' },
      { Name: 'email', Value: 'invited@example.com' }]);

    await createUserInvite(mockExperimentId, mockInvitedUserEmail, mockRole, mockInviterUser);

    expect(mockUserAccess.grantAccess).toHaveBeenCalledWith('mock-user-id', mockExperimentId, mockRole);
    expect(mockUserAccess.grantAccess).toHaveBeenCalledTimes(1);

    expect(sendEmail).toHaveBeenCalledTimes(1);

    const emailBody = sendEmail.mock.calls[0][0];

    expect(emailBody).toMatchSnapshot();
  });

  it('Sends an invitation to sign up if user is not registered', async () => {
    getAwsUserAttributes.mockImplementationOnce(() => {
      const error = Object.assign(new Error('User not found'), { code: 'UserNotFoundException' });
      throw error;
    });

    await createUserInvite(mockExperimentId, mockInvitedUserEmail, mockRole, mockInviterUser);

    expect(mockUserAccess.addToInviteAccess).toHaveBeenCalledWith(mockInvitedUserEmail, mockExperimentId, mockRole);
    expect(mockUserAccess.addToInviteAccess).toHaveBeenCalledTimes(1);

    expect(sendEmail).toHaveBeenCalledTimes(1);

    const emailBody = sendEmail.mock.calls[0][0];

    expect(emailBody).toMatchSnapshot();
  });

  it('Throws error if there is an error when trying to check for existence of invited user', async () => {
    getAwsUserAttributes.mockImplementationOnce(() => Promise.reject(new Error('Some error')));

    expect(async () => {
      await createUserInvite(mockExperimentId);
    }).rejects.toThrow();

    expect(mockUserAccess.grantAccess).not.toHaveBeenCalled();
    expect(mockUserAccess.addToInviteAccess).not.toHaveBeenCalled();
    expect(sendEmail).not.toHaveBeenCalled();
  });
});
