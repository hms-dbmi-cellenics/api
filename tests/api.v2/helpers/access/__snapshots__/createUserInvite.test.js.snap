// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creatUserInvite Grants access if user is already registered and sends an email 1`] = `
Object {
  "Destination": Object {
    "ToAddresses": Array [
      "invited@example.com",
    ],
  },
  "Message": Object {
    "Body": Object {
      "Html": Object {
        "Charset": "UTF-8",
        "Data": "
        <html>
        <head>
        </head>
        <body>
            <p>
              Hello, <br/>
              Your collaborator inviter@example.com has invited you to explore their single cell RNA-seq project in Cellenics. <br/><br/>
              Access it with the link below: <br/>
              <a href=\\"https://test.com/experiments/experimentId/data-exploration\\">https://test.com/experiments/experimentId/data-exploration</a> <br/> <br/>
              Cellenics is a user-friendly online tool for single cell RNA-seq data analysis. <br/>
              The platform is designed specifically for biologists, and it's completely free for academic researchers.<br/><br/><div>test hi info</div>
              Best Regards, <br/>
              Cellenics team
            <p/>
        </body>
      </html>",
      },
      "Text": Object {
        "Charset": "UTF-8",
        "Data": "
        <html>
        <head>
        </head>
        <body>
            <p>
              Hello, <br/>
              Your collaborator inviter@example.com has invited you to explore their single cell RNA-seq project in Cellenics. <br/><br/>
              Access it with the link below: <br/>
              <a href=\\"https://test.com/experiments/experimentId/data-exploration\\">https://test.com/experiments/experimentId/data-exploration</a> <br/> <br/>
              Cellenics is a user-friendly online tool for single cell RNA-seq data analysis. <br/>
              The platform is designed specifically for biologists, and it's completely free for academic researchers.<br/><br/><div>test hi info</div>
              Best Regards, <br/>
              Cellenics team
            <p/>
        </body>
      </html>",
      },
    },
    "Subject": Object {
      "Charset": "UTF-8",
      "Data": "Invitation to join a project in Cellenics",
    },
  },
  "Source": "some@email.com",
}
`;

exports[`creatUserInvite Sends an invitation to sign up if user is not registered 1`] = `
Object {
  "Destination": Object {
    "ToAddresses": Array [
      "invited@example.com",
    ],
  },
  "Message": Object {
    "Body": Object {
      "Html": Object {
        "Charset": "UTF-8",
        "Data": "
        <html>
        <head>
        </head>
        <body>
            <p>
              Hello, <br/>
              Your collaborator inviter@example.com has invited you to explore their single cell RNA-seq project in Cellenics. <br/><br/>
              You need to create a Cellenics account in order to access the project. <br/>
              Register at <a href=\\"https://test.com\\">https://test.com</a> using this email address.<br/> <br/>

              Cellenics is a user-friendly online tool for single cell RNA-seq data analysis. <br/>
              The platform is designed specifically for biologists, and it's completely free for academic researchers.<br/><br/><div>test hi info</div>
              Best Regards, <br/>
              Cellenics team
            <p/>
        </body>
      </html>",
      },
      "Text": Object {
        "Charset": "UTF-8",
        "Data": "
        <html>
        <head>
        </head>
        <body>
            <p>
              Hello, <br/>
              Your collaborator inviter@example.com has invited you to explore their single cell RNA-seq project in Cellenics. <br/><br/>
              You need to create a Cellenics account in order to access the project. <br/>
              Register at <a href=\\"https://test.com\\">https://test.com</a> using this email address.<br/> <br/>

              Cellenics is a user-friendly online tool for single cell RNA-seq data analysis. <br/>
              The platform is designed specifically for biologists, and it's completely free for academic researchers.<br/><br/><div>test hi info</div>
              Best Regards, <br/>
              Cellenics team
            <p/>
        </body>
      </html>",
      },
    },
    "Subject": Object {
      "Charset": "UTF-8",
      "Data": "Invitation to join a project in Cellenics",
    },
  },
  "Source": "some@email.com",
}
`;
